core/src/main/scala/org/allenai/common/SeekableSource.scala

```diff
@@ -0,0 +1,256 @@
+  // passing buffers of size 1, 2, 3, etc. until it successfully decodes the next character) are
+  // incredibly inefficient.
+  //
+  // This uses manual decoding of individual characters for the Iterator[Char], but uses builtin
+  // decoding for linewise reading for the Iterator[String] returned from getLines().
+
+  /** True if we're decoding as UTF-8; false if we're decoding as ISO-8859-1. */
+  val useUtf8 = if (codec.name == Codec.UTF8.name) {
+    true
+  } else if (codec.name == Codec.ISO8859.name) {
+    false
+  } else {
+    throw new IllegalArgumentException(s"Unsupported codec ${codec.name}")
+  }
+
+  /** True if the inFile still has data to read. */
+  private[common] var inputRemaining = true
+
+  /** The buffer to read input into and out of. Initialized to empty. */
+  private[common] var inBuffer = {
+    val buffer = ByteBuffer.allocateDirect(bufferSize)
+    buffer.limit(0)
+    buffer
+  }
+
+  /** Returns true if the source has more input, populating the input buffer if need be. */
+  override def hasNext: Boolean = inBuffer.hasRemaining || (inputRemaining && fillBuffer())
+
+  /** Reads and returns a single character from a the file. */
+  override def next(): Char = {
+    if (inBuffer.hasRemaining || (inputRemaining && fillBuffer())) {
+      if (useUtf8) {
+        nextUtf8
+      } else {
+        nextIso8859
+      }
+    } else {
+      throw new NoSuchElementException("next() called on empty SeekableSource")
+    }
+  }
+
+  /** Positions this source at the given byte offset in the input file. Note that this will allow
+    * you to position past the end of the file, in which case no input will be read.
+    */
+  def position(newPosition: Long): Unit = {
+    inFile.position(newPosition)
+    inBuffer.limit(0)
+    inputRemaining = true
+  }
+
+  /** @return the current position in the file */
+  def position: Long = inFile.position - inBuffer.remaining
+
+  /** @return the next character in the buffer, decoded from ISO-8859-1 */
+  protected[common] def nextIso8859: Char = {
+    // `toChar` doesn't work as expected with bytes; they're first widened with sign extension,
+    // meaning that bytes with the high bit set expand into two-byte characters with the high byte
+    // set to 0xff. Mask into an int first to fix this.
+    (inBuffer.get() & 0xff).toChar
+  }
+
+  /** @return the next character in the buffer, decoded from UTF-8 */
+  protected[common] def nextUtf8: Char = {
+    // Ensure we are either at the end of file, or have at least three bytes that are readable.
+    if (inBuffer.remaining() < 3 && inputRemaining) {
+      fillBuffer()
+    }
+    // See https://en.wikipedia.org/wiki/UTF-8 for details. This is checking the first byte for the
+    // total length of the encoding.
+    val first = inBuffer.get()
+    val intVal = if ((first & 0x80) == 0) {
+      return first.toChar
+    } else if ((first & 0xe0) == 0xc0) {
```

Comment:
This is conditional of the byte starting with `110`. Is that what you want? It's not what the comment says below.